protocol/bc/txinput.go

```diff
@@ -18,279 +18,233 @@ type (
-					return err
-				}
-
-			default:
-				return fmt.Errorf("unsupported input type %d", icType[0])
+				t.TypedInput = ii
+			}
+		case 1:
+			inp := new(SpendInput)
+			err = inp.readCommitment(r, txVersion, t.AssetVersion)
+			if err != nil {
+				return errors.Wrap(err, "reading spend input commitment")
 			}
+			t.TypedInput = inp
+		default:
+			return fmt.Errorf("unsupported input type %d", icType[0])
 		}
 		return nil
 	})
 	if err != nil {
-		return err
+		return errors.Wrap(err, "reading input commitment")
 	}
 
 	t.ReferenceData, _, err = blockchain.ReadVarstr31(r)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "reading input reference data")
 	}
 
 	_, err = blockchain.ReadExtensibleString(r, false, func(r io.Reader) error {
 		// TODO(bobg): test that serialization flags include SerWitness, when we relax the serflags-must-be-0x7 rule
-		if ii != nil {
-			// read IssuanceInput witness
-			_, err = io.ReadFull(r, ii.InitialBlock[:])
-			if err != nil {
-				return err
-			}
-
-			ii.AssetDefinition, _, err = blockchain.ReadVarstr31(r)
-			if err != nil {
-				return err
-			}
-
-			ii.VMVersion, _, err = blockchain.ReadVarint63(r)
-			if err != nil {
-				return err
-			}
-
-			ii.IssuanceProgram, _, err = blockchain.ReadVarstr31(r)
-			if err != nil {
-				return err
-			}
-
-			computedAssetID := ComputeAssetID(ii.IssuanceProgram, ii.InitialBlock, ii.VMVersion, ii.AssetDefinitionHash())
-			if computedAssetID != assetID {
-				return errBadAssetID
-			}
-		}
-		args, _, err := blockchain.ReadVarstrList(r)
-		if err != nil {
-			return err
-		}
-		if ii != nil {
-			ii.Arguments = args
-		} else if si != nil {
-			si.Arguments = args
-		}
-		return nil
+		return t.TypedInput.readWitness(r, t.AssetVersion)
 	})
 	if err != nil {
-		return err
+		return errors.Wrap(err, "reading input witness")
 	}
+
 	if ii != nil {
-		t.TypedInput = ii
-	} else if si != nil {
-		t.TypedInput = si
+		if assetID != ii.AssetID() {
+			return errBadAssetID
+		}
 	}
+
 	return nil
 }
 
-func (t *TxInput) writeTo(w io.Writer, serflags uint8) error {
-	_, err := blockchain.WriteVarint63(w, t.AssetVersion)
-	if err != nil {
-		return errors.Wrap(err, "writing asset version")
-	}
+func (t *TxInput) AssetAmount() (assetAmount AssetAmount) {
+	switch inp := t.TypedInput.(type) {
+	case *IssuanceInput:
+		assetAmount.AssetID = inp.AssetID()
+		assetAmount.Amount = inp.Amount
+		return assetAmount
 
-	_, err = blockchain.WriteExtensibleString(w, t.WriteInputCommitment)
-	if err != nil {
-		return errors.Wrap(err, "writing input commitment")
+	case *SpendInput:
+		return inp.AssetAmount
 	}
+	return assetAmount
+}
 
-	_, err = blockchain.WriteVarstr31(w, t.ReferenceData)
-	if err != nil {
-		return errors.Wrap(err, "writing reference data")
+func (t *TxInput) AssetID() (assetID AssetID) {
+	switch inp := t.TypedInput.(type) {
+	case *IssuanceInput:
+		return inp.AssetID()
+
+	case *SpendInput:
+		return inp.AssetID
 	}
+	return assetID
```

Comment:
Never mind, I just realized the return values are named. I'm still waking up today.